src/frontend/src/scheduler/plan_fragmenter.rs

```diff
@@ -418,6 +514,94 @@ impl StageGraph {
 
         ret.into_iter().rev()
     }
+
+    pub async fn complete(
+        self,
+        catalog_reader: &CatalogReader,
+        worker_node_manager: &WorkerNodeManagerRef,
+    ) -> SchedulerResult<StageGraph> {
+        let mut complete_stages = HashMap::new();
+        self.complete_stage(
+            self.stages.get(&self.root_stage_id).unwrap().clone(),
+            None,
+            &mut complete_stages,
+            catalog_reader,
+            worker_node_manager,
+        )
+        .await?;
+        Ok(StageGraph {
+            root_stage_id: self.root_stage_id,
+            stages: complete_stages,
+            child_edges: self.child_edges,
+            parent_edges: self.parent_edges,
+        })
+    }
+
+    #[async_recursion]
+    async fn complete_stage(
+        &self,
+        stage: QueryStageRef,
+        exchange_info: Option<ExchangeInfo>,
+        complete_stages: &mut HashMap<StageId, QueryStageRef>,
+        catalog_reader: &CatalogReader,
+        worker_node_manager: &WorkerNodeManagerRef,
+    ) -> SchedulerResult<()> {
+        let parallelism = if stage.parallelism.is_some() {
```

Comment:
If parallelism is not `None`, then it means it's not changed. The logic is correct here, but please add some comment to explain it.